### PR TITLE
build: add check for functional environment

### PR DESCRIPTION
**Pull request by nixpanic**
In case building fails, users may not always understand what is wrong. By adding a `check-env.sh` script, most common issues should be caught and explained.

### CODE REVIEW

File: scripts/check-env.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/bin/sh
```
Comment (Madhu-1) [inline scripts/check-env.sh:1]: we need to check `go` is installed or not in the first place

File: scripts/check-env.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+# Golang needs to be > 1.13
+GO_MAJOR=$(go_version 1)
+GO_MINOR=$(go_version 2)
+if ! [ ${GO_MAJOR} -gt 1 ]
+then
+	if ! ( [ ${GO_MAJOR} -eq 1 ] && [ ${GO_MINOR} -ge 13 ] )
+	then
+		fail "go version needs to be >= 1.13"
```
Comment (Madhu-1) [inline scripts/check-env.sh:29]: can we just exit here, no need to perform other checks as the go version itself is not compatible

File: scripts/check-env.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	fi
+fi
+
+# we're building with modules, so GO11MODULE needs to be 'on'
+[ "$(go env GO111MODULE)" = 'on' ] || fail "GO111MODULE should be set to 'on'"
```
Comment (Madhu-1) [inline scripts/check-env.sh:34]: check go111MODULES is off here as it can be auto also

File: scripts/check-env.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+# CGO needs to be enabled, we build with go-ceph language bindings
+[ "$(go env CGO_ENABLED)" = '1' ] || fail "CGO_ENABLED should be set to '1'"
```
Comment (Madhu-1) [inline scripts/check-env.sh:37]: the comparison seems wrong here

File: scripts/check-env.sh (hunk 2 above)
Comment (nixpanic) [reply scripts/check-env.sh:29]: It is useful to list all failures at the end. Users can then address multiple issues before trying again.

File: scripts/check-env.sh (hunk 3 above)
Comment (nixpanic) [reply scripts/check-env.sh:34]: good point!

File: scripts/check-env.sh (hunk 1 above)
Comment (nixpanic) [reply scripts/check-env.sh:1]: Will add that

File: scripts/check-env.sh (hunk 4 above)
Comment (nixpanic) [reply scripts/check-env.sh]: What is wrong about it? It needs to be `1` OR needs to fail.

Comment (Madhu-1) [reply scripts/check-env.sh]: it should be `==` or `=` here?

Comment (nixpanic) [reply scripts/check-env.sh]: No, `=` is correct for shell scripts.

File: deploy/cephcsi/image/Dockerfile [hunk 5]
```diff
@@ -21,6 +21,7 @@ RUN dnf install libcephfs-devel librados-devel librbd-devel /usr/bin/cc make -y
 
 ENV GOROOT=${GOROOT} \
     GOPATH=/go \
+    GO111MODULE=on \
```
Comment (Madhu-1) [inline deploy/cephcsi/image/Dockerfile:24]: GOMODULE is on by default in go 1.13.x we don't need to set this explicitly

Comment (nixpanic) [reply deploy/cephcsi/image/Dockerfile:24]: What is the reason then that building fails? `go env` does not return `on` or `auto` it seems? Other variables that are enabled by default return a non-empty value with `go env`.

Maybe I am missing something?

Comment (nixpanic) [reply deploy/cephcsi/image/Dockerfile:24]: ok, I see it is `on` by default, but `go env GO111MODULE` returns `on` in that case too. By setting the variable empty on the cmdline `make GO111MODULE=""` the script detects it as `on`. By setting it to `off` or an other value, the script reports the error.

Comment (Madhu-1) [reply deploy/cephcsi/image/Dockerfile:24]: `""` is also a valid value. we need to consider it

Comment (nixpanic) [reply deploy/cephcsi/image/Dockerfile:24]: Depending on the Golang version, `go env GO111MODULE` may return an empty string... Which means `on` or `auto`, so an empty-non-error-return should be valid too.

File: scripts/check-env.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	if ! { [ "${GO_MAJOR}" -eq 1 ] && [ "${GO_MINOR}" -ge 13 ]; }
+	then
+		fail "go version needs to be >= 1.13"
+	fi
+fi
+
+# we're building with modules, so GO111MODULE needs to be 'on' or 'auto'
+# some versions of Go return nothing when GO111MODULE is not explicitly
+# configured
+[ "$(go env GO111MODULE 2>&1)" = '' ] || [ "$(go env GO111MODULE)" = 'on' ] || [ "$(go env GO111MODULE)" = 'auto' ] \
```
Comment (Madhu-1) [inline scripts/check-env.sh:39]: can this be simplified by just checking the GO111MODULE  is off?

Comment (nixpanic) [reply scripts/check-env.sh:39]: maybe, but we would also need to check if it is set to an invalid value in that case, I think? that would not simplify things much more

**Comment by Madhu-1** [comment]
@humblec PTAL

**Comment by humblec** [comment]
@nixpanic can you also document when this makefile has to be updated in the developer guide ?

**Comment by nixpanic** [comment]
> @nixpanic can you also document when this makefile has to be updated in the developer guide ?

I'm not sure what you mean. Why would the makefile need to be updated?

**Comment by humblec** [comment]
> > @nixpanic can you also document when this makefile has to be updated in the developer guide ?
> 
> I'm not sure what you mean. Why would the makefile need to be updated?

@nixpanic we check the version and fail in makefile target https://github.com/ceph/ceph-csi/pull/1005/files#diff-212af17ce32e5c0911abd264ca05d9fdR30 . What I was asking here is, when can we(developer) change the `go` version compatibility check to a different version and make a change in the make `target` ?

Approving this pr for now though.

**Comment by nixpanic** [comment]
On Mon, May 11, 2020 at 11:35:25AM -0700, Humble Devassy Chirammal wrote:
> > > @nixpanic can you also document when this makefile has to be updated in the developer guide ?
> > 
> > I'm not sure what you mean. Why would the makefile need to be updated?
> 
> @nixpanic we check the version and fail in makefile target https://github.com/ceph/ceph-csi/pull/1005/files#diff-212af17ce32e5c0911abd264ca05d9fdR30 . What I was asking here is, when can we(developer) change the `go` version compatibility check to a different version and make a change in the make `target` ?

In case a newer version of Golang is needed, the scripts/check-env.sh
can be updated with the new requirement(s). This is not different for
Golang versions, or any other settings or tools. Everything that is
needed to build Ceph-CSI is now in the scripts/check-env.sh script.

We probably should add a check for libcephfs, librados and librbd header
files too. But that is a little more tricky to get right.

**Comment by nixpanic** [comment]
@mergifyio refresh

**Comment by nixpanic** [comment]
@mergifyio rebase

**Pull request merged**